### PR TITLE
Use TreeBacked<BeaconState> everywhere in lodestar package

### LINKED ISSUE #1100: Use TreeBacked<BeaconState> in IBeaconChain
_Originally posted by @wemeetagain in https://github.com/_render_node/MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjc5ODc3NTgxOnYy/pull_request_review_threads/discussion_

I think we should create a:
```typescript
export interface ITreeStateContext {
  state: TreeBacked<BeaconState>;
  epochCtx: EpochContext;
}
```
and use TreeBacked beacon state here and across the `lodestar` package, rather than currently at the boundary of `StateContextCache`.

eg:
```typescript
interface IBeaconChain {
  ...
  getHeadStateContext(): Promise<ITreeStateContext | null>;
  getHeadState(): Promise<TreeBacked<BeaconState> | null>;
}
```

Rationale:

We rely on the efficiency of a tree-backed state, and, practically, can't run a production beacon node without it. If we accept this, there's not a benefit to obscuring the backing throughout the `lodestar` package, as we have been doing.

This will also open the way for serving proofs :)

### PR DESCRIPTION
**Pull request by twoeths**
resolves #1100 
+ Fix unit test interopSubnetsJoiningTask.test.ts

### CODE REVIEW

File: packages/lodestar/test/utils/state.ts [hunk 1]
```diff
@@ -61,6 +66,12 @@ export function generateState(opts?: TestBeaconState, config = mainnetConfig): B
       epoch: GENESIS_EPOCH,
       root: ZERO_HASH,
     },
-    ...opts,
   };
+  treeBackedState = treeBackedState || config.types.BeaconState.tree.createValue(defaultState);
```
Comment (twoeths) [inline packages/lodestar/test/utils/state.ts:70]: Generating tree backed BeaconState is expensive (~4s in my environment) so I made a default tree backed state here. For the 1st test calling this function, I have to increase timeout for it. The down side is if we run 1 single test in local environment using `it.only` we may see this timeout issue if that test calls `generateState()`

Not sure if there're any better ways to resolve this.

Comment (wemeetagain) [reply packages/lodestar/test/utils/state.ts:70]: Maybe we can add a mocha `before` (outside of any `describe`) to call generateState
eg: `test/unit/setupState.ts` and the same for e2e

Comment (twoeths) [reply packages/lodestar/test/utils/state.ts:70]: thanks, that's a good idea

File: packages/lodestar/test/utils/state.ts [hunk 2]
```diff
@@ -2,21 +2,26 @@ import {BeaconState} from "@chainsafe/lodestar-types";
 import {GENESIS_EPOCH, GENESIS_SLOT, ZERO_HASH} from "../../src/constants";
 import {generateEmptyBlock} from "./block";
 import {config as mainnetConfig} from "@chainsafe/lodestar-config/lib/presets/mainnet";
```
Comment (mpetrunic) [inline packages/lodestar/test/utils/state.ts]:
```suggestion
import {config as minimal} from "@chainsafe/lodestar-config/lib/presets/minimal";
```

File: packages/lodestar/test/utils/state.ts (hunk 1 above)
Comment (mpetrunic) [reply packages/lodestar/test/utils/state.ts:70]: Please use minimal preset for state, we cannot have mainnet state in unit tests

**Review by mpetrunic — CHANGES_REQUESTED** [review]

File: packages/lodestar/test/utils/state.ts (hunk 1 above)
Comment (twoeths) [reply packages/lodestar/test/utils/state.ts:70]: so we should use minimal preset for unit tests and mainnet preset for e2e tests? Fyi, this file is shared for both unit and e2e tests.

Comment (mpetrunic) [reply packages/lodestar/test/utils/state.ts:70]: In my opinion we can use minimal in e2e as well. Only place for mainnet would be benchmark

**Review by mpetrunic — APPROVED** [review]

**Review by wemeetagain — APPROVED** [review]

**Pull request merged**